src/dataflow-types/src/logging.rs

```diff
@@ -56,6 +57,10 @@ pub enum MaterializedLog {
     PeekDuration,
     SourceInfo,
 }
+#[derive(Hash, Eq, PartialEq, Debug, Clone, Serialize, Deserialize)]
+pub enum RDKafkaLog {
+    ConsumerStatistics,
+}
```

Comment:
I think this probably just wants to a a `KafkaConsumerStatistics` variant in the existing `Materialized` enum! The existing separation for timely and differential logs is because those logs integrate with messages that are already integrated with the timely logging framework. As far as the timely logger is concerned, events generated by librdkafka look identical to events generated by other parts of Materialize, so I think you'll save yourself some trouble if you just combine the two.

Reply:
Thanks for the feedback! I'm not sure if you saw the thread on Slack yesterday -- I originally tried adding this to the `Materialized` enum for the reasons you stated above. Unfortunately, `rdkafka` expects the `GlueConsumerContext` to be `Send + Sync`, prohibiting the ability to pass the `Logger` object into the callback's context. Frank advised that I write a new `MPSC Logger` , as I'll need to implement one that is both `Send + Sync` safe. By putting this into it's own enum, we won't need to modify the `Materialized Logger` for messages that it will never see.

Do you have thoughts on this? I'd like to keep the implementation simple and coherent with the rest of the codebase. If there's a way to use the `Materialized Logger` in the `GlueConsumerContext`, I'd be very happy to do that.

Reply:
Oh, hm, I see. I think I understand what Frank is proposing, and it seems nice and general purpose, but it seems like it will require trudging through quite a bit of pain to implement. Personally I'd do something like this:

```diff
diff --git a/src/dataflow/src/source/kafka.rs b/src/dataflow/src/source/kafka.rs
index cc567dd71..61419acc1 100644
--- a/src/dataflow/src/source/kafka.rs
+++ b/src/dataflow/src/source/kafka.rs
@@ -62,6 +62,8 @@ pub struct KafkaSourceInfo {
     cached_files: Vec<PathBuf>,
     /// Timely worker logger for source events
     logger: Option<Logger>,
+    /// Receiver for librdkafka statistics events.
+    stats_rx: crossbeam_channel::Receiver<Statistics>,
 }
 
 impl SourceConstructor<Vec<u8>> for KafkaSourceInfo {
@@ -157,6 +159,12 @@ impl SourceInfo<Vec<u8>> for KafkaSourceInfo {
             }
         }
 
+        if let Some(logger) = self.logger {
+            while let Ok(statistics) = self.stats_rx.try_recv() {
+                logger.log(MaterializedEvent::KafkaConsumerStatistics { /* whatever */});
+            }
+        }
+
         let mut next_message = NextMessage::Pending;
         let consumer_count = self.get_partition_consumers_count();
         let mut attempts = 0;
@@ -329,8 +337,12 @@ impl KafkaSourceInfo {
             cluster_id,
             &config_options,
         );
+        let (stats_tx, stats_rx) = crossbeam_channel::unbounded();
         let consumer: BaseConsumer<GlueConsumerContext> = kafka_config
-            .create_with_context(GlueConsumerContext(consumer_activator))
+            .create_with_context(GlueConsumerContext {
+                activator: consumer_activator,
+                stats_tx,
+            })
             .expect("Failed to create Kafka Consumer");
         let cached_files = kc
             .cached_files
@@ -621,17 +633,21 @@ impl PartitionConsumer {
 
 /// An implementation of [`ConsumerContext`] that unparks the wrapped thread
 /// when the message queue switches from nonempty to empty.
-struct GlueConsumerContext(SyncActivator);
+struct GlueConsumerContext {
+    activator: SyncActivator,
+    stats_tx: crossbeam_channel::Sender<Statistics>,
+}
 
 impl ClientContext for GlueConsumerContext {
     fn stats(&self, statistics: Statistics) {
-        info!("Client stats: {:#?}", statistics);
+        self.stats_tx.send(statistics).expect("timely operator hung up while Kafka source active");
+        self.activate();
     }
 }
 
 impl GlueConsumerContext {
     fn activate(&self) {
-        self.0
+        self.activator
             .activate()
             .expect("timely operator hung up while Kafka source active");
     }

```

Possibly that's a terrible idea for some reason, but seems easy enough just to try out.

Reply:
That does seem much easier than the route that I was going down -- thank you! I'll give this a shot first thing tomorrow!

Reply:
mind linking the slack thread, @cirego  ?

Reply:
It was in the Rust channel. Here's the link to Frank's comment about creating an `Arc<EventLinkArc>` class: https://materializeinc.slack.com/archives/CMH6PG4CW/p1615938814067100